pkg/storegateway/series_chunks.go

```diff
@@ -331,42 +386,248 @@ func (c *loadingSeriesChunksSetIterator) Next() (retHasNext bool) {
 		}
 	}()
 
-	// The series slice is guaranteed to have at least the requested capacity,
-	// so can safely expand it.
-	nextSet.series = nextSet.series[:nextUnloaded.len()]
-
-	c.chunkReaders.reset()
+	partialSeries := make([]partialSeriesChunksSet, len(nextUnloaded.series))
 
+	totalGroups := 0
 	for i, s := range nextUnloaded.series {
-		nextSet.series[i].lset = s.lset
-		nextSet.series[i].chks = nextSet.newSeriesAggrChunkSlice(len(s.chunks))
+		numGroups := len(s.groups)
+		totalGroups += numGroups
+
+		partialSeries[i].groups = s.groups
+		partialSeries[i].rawGroups = make([][]byte, numGroups)
+		chunksCount := 0
+		for _, g := range s.groups {
+			chunksCount += len(g.chunks)
+		}
+		partialSeries[i].parsedChunks = nextSet.newSeriesAggrChunkSlice(chunksCount)
+	}
 
-		for j, chunk := range s.chunks {
-			nextSet.series[i].chks[j].MinTime = chunk.minTime
-			nextSet.series[i].chks[j].MaxTime = chunk.maxTime
+	// Create a batched memory pool that can be released all at once. We keep all chunks bytes there.
+	var cachedRanges map[chunkscache.Range][]byte
+	if c.cache != nil {
+		cachedRanges, _ = c.cache.FetchMultiChunks(c.ctx, c.userID, toCacheRanges(nextUnloaded.series, totalGroups), chunksPool)
```

Comment:
[nit] If we don't need `misses` returned by `FetchMultiChunks()` then we can remove it from return value. Maybe we can save some bytes.